grinder/core.py

```diff
@@ -1167,10 +1324,72 @@ def nmap_scan(
             self.shodan_processed_results[host]["nmap_scan"] = nmap_results.get(host)
         for host in self.censys_processed_results.keys():
             self.censys_processed_results[host]["nmap_scan"] = nmap_results.get(host)
+        for host in self.masscan_results.keys():
+            self.masscan_results[host]["nmap_scan"] = nmap_results.get(host)
         # Trigger to update overall results (shodan + censys as combined results)
         for host in self.combined_results.keys():
             self.combined_results[host]["nmap_scan"] = nmap_results.get(host)
 
+
+    @timer
+    @exception_handler(expected_exception=GrinderCoreNmapScanError)
+    def masscan_scan(
+        self,
+        hosts: str = None,
+        ports: str = DefaultMasscanScanValues.PORTS,
+        top_ports: int = DefaultMasscanScanValues.TOP_PORTS,
+        rate: int = DefaultMasscanScanValues.RATE,
+        arguments: str = DefaultMasscanScanValues.ARGUMENTS,
+        sudo: bool = DefaultMasscanScanValues.SUDO,
+    ) -> dict:
+        """
+        Initiate Masscan scan on hosts
+
+        :param hosts: ip to scan
+        :param ports: ports to scan
+        :param top_ports: number of first ports in top ports list
+        :param rate: packet rate
+        :param arguments: masscan arguments
+        :param sudo: sudo if needed
+        :return: None
+        """
+        print(
+            f'│ Masscan scan arguments: {arguments or None}, rate "{str(rate)}", '
+            f'hosts: "{str(hosts)}", ports: "{str(ports)}", top ports: "{str(top_ports)}"',
+        )
+
+        # check if we already root user,
+        # this is necessary to run grinder in docker
+        if os.getuid() == 0:
+            sudo = False
+
+        if ports != "" and top_ports != 0:
+            ports += ","
+        ports += ",".join([str(port) for port in TopPorts.MASSCAN_TOP_PORTS[:top_ports]])
```

Comment:
<img width="1434" alt="bug-3" src="https://user-images.githubusercontent.com/24269213/85234608-c2f18d80-b438-11ea-963e-024649f09dd2.png">

```
        "masscan_settings": [
            {
                "hosts": "87.250.250.96/27",
                "ports": "",
                "rate": "1000",
                "top-ports": "10"
            }
        ],
```

Reply:
Just discovered ports were not in first 10 ports of top-port list, try use 20 instead of 10.